spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,384 @@
+
+#### Permissioning
+
+- Channel ends should be permissioned to one module on each end.
+
+## Technical Specification
+
+![channel-state-machine](channel-state-machine.png)
+
+### Channel opening handshake
+
+```typescript
+interface ChanOpenInit {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  counterpartyChannelIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+}
+```
+
+```typescript
+function chanOpenInit() {
+  moduleIdentifier = getCallingModule()
+  assert(get("connections/{connectionIdentifier}/channels/{channelIdentifier}") === null)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  channel = Channel{INIT, moduleIdentifier, counterpartyModuleIdentifier, counterpartyChannelIdentifier, 0, 0}
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+```typescript
+interface ChanOpenTry {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  counterpartyChannelIdentifier: Identifier
+  moduleIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+  proofInit: CommitmentProof
+}
+```
+
+```typescript
+function chanOpenTry() {
+  assert(get("connections/{connectionIdentifier}/channels/{channelIdentifier}") === null)
+  assert(getCallingModule() === moduleIdentifier)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(verifyMembership(
+    consensusState,
+    proofInit,
+    "connections/{connection.counterpartyConnectionIdentifier}/channels/{counterpartyChannelIdentifier}",
+    Channel{INIT, counterpartyModuleIdentifier, moduleIdentifier, channelIdentifier, 0, 0}
+  ))
+  channel = Channel{TRYOPEN, moduleIdentifier, counterpartyModuleIdentifier, counterpartyChannelIdentifier, 0, 0}
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+```typescript
+interface ChanOpenAck {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  proofTry: CommitmentProof
+}
+```
+
+```typescript
+function chanOpenAck() {
+  channel = get("connections/{connectionIdentifier}/channels/{channelIdentifier}")
+  assert(channel.state === INIT)
+  assert(getCallingModule() === channel.moduleIdentifier)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(verifyMembership(
+    consensusState.getRoot(),
+    proofTry,
+    "connections/{connection.counterpartyConnectionIdentifier}/channels/{channel.counterpartyChannelIdentifier}",
+    Channel{TRYOPEN, channel.counterpartyModuleIdentifier, channel.moduleIdentifier, channelIdentifier, 0, 0}
```

Comment:
Not sure what the function does but here it looks like we pass the channel endpoint at the other end?

Reply:
Fixed.